sharding/p2p/discovery.go

```diff
@@ -0,0 +1,71 @@
+package p2p
+
+import (
+	"context"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ps "github.com/libp2p/go-libp2p-peerstore"
+	mdns "github.com/libp2p/go-libp2p/p2p/discovery"
+	pb "github.com/prysmaticlabs/geth-sharding/sharding/p2p/proto/v1"
+	log "github.com/sirupsen/logrus"
+)
+
+// Discovery interval for multicast DNS querying.
+var discoveryInterval = 1 * time.Minute
+
+// mDNSTag is the name of the mDNS service.
+var mDNSTag = mdns.ServiceTag
+
+// startDiscovery protocols. Currently, this supports discovery via multicast
+// DNS peer discovery.
+//
+// TODO: add other discovery protocols such as DHT, etc.
+func startDiscovery(ctx context.Context, host host.Host, gsub topicPeerLister) error {
+	mdnsService, err := mdns.NewMdnsService(ctx, host, discoveryInterval, mDNSTag)
+	if err != nil {
+		return err
+	}
+
+	mdnsService.RegisterNotifee(&discovery{ctx, host, gsub})
+
+	return nil
+}
+
+// topicPeerLister has a method to return connected peers on a given topic.
+// This is implemented by floodsub.PubSub.
+type topicPeerLister interface {
+	ListPeers(string) []peer.ID
+}
+
+// Discovery implements mDNS notifee interface.
+type discovery struct {
+	ctx  context.Context
+	host host.Host
+
+	// Required for helper method.
+	gsub topicPeerLister
+}
+
+// HandlePeerFound registers the peer with the host.
+func (d *discovery) HandlePeerFound(pi ps.PeerInfo) {
+	d.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, ps.PermanentAddrTTL)
+	if err := d.host.Connect(d.ctx, pi); err != nil {
+		log.Warnf("Failed to connect to peer: %v", err)
```

Comment:
Wouldn't error be more appropriate? 
Warning is like "you should probably take a look at this"
Error is like "something failed but I'm not quitting"

Reply:
I'm not sure it's really error level. Imagine that they discover a peer but then the peer goes offline or refuses the connection. 

In geth, these are debug level logs so warning might even be too verbose.